File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.easy.dd2d

import better.files.File
import nl.knaw.dans.easy.dd2d.OutboxSubdir.{ FAILED, OutboxSubdir, PROCESSED, REJECTED }
import nl.knaw.dans.easy.dd2d.dansbag.InformationPackageType.InformationPackageType
import nl.knaw.dans.easy.dd2d.dansbag.{ DansBagValidationResult, DansBagValidator, InformationPackageType }
import nl.knaw.dans.easy.dd2d.mapping.FieldMap
import nl.knaw.dans.lib.dataverse.DataverseClient
import nl.knaw.dans.lib.dataverse.model.dataset
import nl.knaw.dans.lib.dataverse.model.dataset.UpdateType.major
import nl.knaw.dans.lib.dataverse.model.dataset.{ Dataset, PrimitiveSingleValueField }
import nl.knaw.dans.lib.error._
import nl.knaw.dans.lib.logging.DebugEnhancedLogging
import nl.knaw.dans.lib.taskqueue.Task
import org.json4s.native.Serialization
import org.json4s.{ DefaultFormats, Formats }

import java.lang.Thread.sleep
import java.net.URI
import java.util.regex.Pattern
import scala.collection.convert.ImplicitConversions.`collection AsScalaIterable`
import scala.collection.mutable.ListBuffer
import scala.language.postfixOps
import scala.util.control.NonFatal
import scala.util.{ Failure, Success, Try }
import scala.xml.{ Elem, Node }

/**
 * Checks one deposit and then ingests it into Dataverse.
 *
 * @param deposit  the deposit to ingest
 */
case class DepositIngestTask(deposit: Deposit,
                             optFileExclusionPattern: Option[Pattern],
                             zipFileHandler: ZipFileHandler,
                             depositorRole: String,
                             deduplicate: Boolean,
                             activeMetadataBlocks: List[String],
                             optDansBagValidator: Option[DansBagValidator],
                             dataverseClient: DataverseClient,
                             publishAwaitUnlockMaxNumberOfRetries: Int,
                             publishAwaitUnlockMillisecondsBetweenRetries: Int,
                             narcisClassification: Elem,
                             iso1ToDataverseLanguage: Map[String, String],
                             iso2ToDataverseLanguage: Map[String, String],
                             variantToLicense: Map[String, String],
                             supportedLicenses: List[URI],
                             repordIdToTerm: Map[String, String],
                             outboxDir: File) extends Task[Deposit] with DebugEnhancedLogging {
  trace(deposit)
  protected val informationPackageType: InformationPackageType = InformationPackageType.SIP
  protected val bagProfileVersion: Int = 1

  private val datasetMetadataMapper = new DepositToDvDatasetMetadataMapper(deduplicate, activeMetadataBlocks, narcisClassification, iso1ToDataverseLanguage, iso2ToDataverseLanguage, repordIdToTerm)
  private val bagDirPath = File(deposit.bagDir.path)

  override def run(): Try[Unit] = {
    doRun()
      .doIfSuccess(_ => {
        logger.info(s"SUCCESS: $deposit")
        deposit.setState("ARCHIVED", "The deposit was successfully ingested in the Data Station and will be automatically archived")
        moveDepositToOutbox(PROCESSED)
      })
      .doIfFailure {
        case e: RejectedDepositException =>
          logger.info(s"REJECTED: $deposit", e)
          deposit.setState("REJECTED", e.msg)
          moveDepositToOutbox(REJECTED)
        case e =>
          logger.info(s"FAILED: $deposit", e)
          deposit.setState("FAILED", e.getMessage)
          moveDepositToOutbox(FAILED)
      }
  }

  private def moveDepositToOutbox(subDir: OutboxSubdir): Unit = {
    try {
      deposit.dir.moveToDirectory(outboxDir / subDir.toString)
    } catch {
      case NonFatal(e) => logger.info(s"Failed to move deposit: $deposit to ${ outboxDir / subDir.toString }", e)
    }
  }

  private def doRun(): Try[Unit] = {
    trace(())
    logger.info(s"Ingesting $deposit into Dataverse")
    for {
      _ <- checkDepositType()
      _ <- validateDeposit()
      dataverseDataset <- getMetadata
      isUpdate <- deposit.isUpdate
      _ = debug(s"isUpdate? = $isUpdate")
      editor = if (isUpdate) newDatasetUpdater(dataverseDataset)
               else newDatasetCreator(dataverseDataset, depositorRole)
      persistentId <- editor.performEdit()
      _ <- publishDataset(persistentId)
      _ <- postPublication(persistentId)
    } yield ()
  }

  protected def checkDepositType(): Try[Unit] = {
    trace(())
    if (deposit.doi.nonEmpty) Failure(new IllegalArgumentException("Deposits must not have an identifier.doi property unless they are migrated"))
    else Success(())
  }

  private def validateDeposit(): Try[Unit] = {
    trace(())
    optDansBagValidator.map {
      dansBagValidator =>
        for {
          validationResult <- dansBagValidator.validateBag(bagDirPath, informationPackageType, bagProfileVersion)
          _ <- rejectIfInvalid(validationResult)
        } yield ()
    }.getOrElse({
      debug("Skipping validation")
      Success(())
    })
  }

  private def rejectIfInvalid(validationResult: DansBagValidationResult): Try[Unit] = Try {
    if (!validationResult.isCompliant) throw RejectedDepositException(deposit,
      s"""
         |Bag was not valid according to Profile Version ${ validationResult.profileVersion }.
         |Violations:
         |${ validationResult.ruleViolations.map(_.map(formatViolation).mkString("\n")).getOrElse("") }
                      """.stripMargin)
  }

  private def formatViolation(v: (String, String)): String = v match {
    case (nr, msg) => s" - [$nr] $msg"
  }

  private def getMetadata: Try[Dataset] = {
    trace(())
    for {
      optDateOfDeposit <- getDateOfDeposit
      datasetContacts <- getDatasetContacts
      ddm <- deposit.tryDdm
      optAgreements <- deposit.tryOptAgreementsXml
      _ <- checkPersonalDataPresent(optAgreements)
      dataverseDataset <- datasetMetadataMapper.toDataverseDataset(ddm, deposit.getOptOtherDoiId, optAgreements, optDateOfDeposit, datasetContacts, deposit.vaultMetadata)
    } yield dataverseDataset
  }

  /*
   * See DD-901. For non-migration imports we will accept missing agreement.xml for now
   */
  protected def checkPersonalDataPresent(optAgreements: Option[Node]): Try[Unit] = {
    Success(())
  }

  protected def getDateOfDeposit: Try[Option[String]] = Try {
    Option.empty
  }

  private def getDatasetContacts: Try[List[FieldMap]] = {
    logger.trace("DepositIngestTask.getDatasetContacts")
    for {
      response <- Try(dataverseClient.admin().listSingleUser(deposit.depositorUserId))
      user <- Try(response.getData)
      datasetContacts <- createDatasetContacts(user.getDisplayName, user.getEmail, Option(user.getAffiliation))
    } yield datasetContacts
  }

  private def createDatasetContacts(name: String, email: String, optAffiliation: Option[String] = None): Try[List[FieldMap]] = Try {
    val subfields = ListBuffer[PrimitiveSingleValueField]()
    subfields.append(new PrimitiveSingleValueField("datasetContactName", name))
    subfields.append(new PrimitiveSingleValueField("datasetContactEmail", email))
    optAffiliation.foreach(affiliation => subfields.append(new PrimitiveSingleValueField("datasetContactAffiliation", affiliation)))

    List(toFieldMap(subfields: _*))
  }

  protected def newDatasetUpdater(dataverseDataset: Dataset): DatasetUpdater = {
    new DatasetUpdater(deposit, optFileExclusionPattern, zipFileHandler, isMigration = false, dataverseDataset.getDatasetVersion.getMetadataBlocks, variantToLicense, supportedLicenses, dataverseClient)
  }

  protected def newDatasetCreator(dataverseDataset: Dataset, depositorRole: String): DatasetCreator = {
    new DatasetCreator(deposit, optFileExclusionPattern, zipFileHandler, depositorRole, isMigration = false, dataverseDataset, variantToLicense, supportedLicenses, dataverseClient)
  }

  protected def publishDataset(persistentId: String): Try[Unit] = {
    trace(persistentId)
    for {
      _ <- Try(dataverseClient.dataset(persistentId).publish(major, true))
      _ <- Try(dataverseClient.dataset(persistentId).awaitUnlock(
        publishAwaitUnlockMaxNumberOfRetries,
        publishAwaitUnlockMillisecondsBetweenRetries))
    } yield ()
  }

  protected def postPublication(persistentId: String): Try[Unit] = {
    trace(persistentId)
    for {
      _ <- waitForReleasedState(persistentId)
      _ <- savePersistentIdentifiersInDepositProperties(persistentId)
    } yield ()
  }

  private def waitForReleasedState(persistentId: String): Try[Unit] = {
    trace(persistentId)
    var numberOfTimesTried = 0

    def getDatasetState: Try[String] = {
      for {
        response <- Try(dataverseClient.dataset(persistentId).getLatestVersion)
        ds <- Try(response.getData)
        state = ds.getLatestVersion.getVersionState
      } yield state
    }

    def slept(): Boolean = {
      debug(s"Sleeping $publishAwaitUnlockMillisecondsBetweenRetries ms before next try..") // TODO: replace with dedicated settings for waiting for pub.
      sleep(publishAwaitUnlockMillisecondsBetweenRetries)
      true
    }

    var maybeState = getDatasetState
    do {
      maybeState = getDatasetState
      numberOfTimesTried += 1
    } while (maybeState.isSuccess && maybeState.get != "RELEASED" && numberOfTimesTried != publishAwaitUnlockMaxNumberOfRetries && slept())

    if (maybeState.isFailure) maybeState.map(_ => ())
    else if (maybeState.get != "RELEASED") Failure(FailedDepositException(deposit, "Dataset did not become RELEASED within the wait period"))
         else Success(())
  }

  private def savePersistentIdentifiersInDepositProperties(persistentId: String): Try[Unit] = {
    logger.trace("ENTER")
    implicit val jsonFormats: Formats = DefaultFormats
    for {
      _ <- Try(dataverseClient.dataset(persistentId).awaitUnlock())
      _ = debug(s"Dataset $persistentId is not locked")
      _ <- deposit.setDoi(persistentId)
      r <- Try(dataverseClient.dataset(persistentId).getVersion)
      _ = if (logger.underlying.isDebugEnabled) debug(Serialization.writePretty(r.getEnvelopeAsJson))
      d <- Try(r.getData)
      v = d.getMetadataBlocks.get("dansDataVaultMetadata")
      optUrn = v.getFields.find("dansNbn" == _.getTypeName)
        .map(_.asInstanceOf[dataset.PrimitiveSingleValueField].getValue)
      _ = if (optUrn.isEmpty) throw new IllegalStateException(s"Dataset $persistentId did not obtain a URN:NBN")
      _ <- deposit.setUrn(optUrn.get)
    } yield ()
  }

  override def getTarget: Deposit = {
    deposit
  }

  override def toString: DepositName = {
    s"DepositIngestTask for $deposit"
  }
}

File: src/main/java/nl/knaw/dans/ingest/core/service/DepositIngestTask.java
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.ingest.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import nl.knaw.dans.ingest.core.TaskEvent;
import nl.knaw.dans.ingest.core.TaskEvent.Result;
import nl.knaw.dans.ingest.core.dataverse.DatasetService;
import nl.knaw.dans.ingest.core.deposit.DepositManager;
import nl.knaw.dans.ingest.core.domain.Deposit;
import nl.knaw.dans.ingest.core.domain.DepositLocation;
import nl.knaw.dans.ingest.core.domain.DepositState;
import nl.knaw.dans.ingest.core.domain.OutboxSubDir;
import nl.knaw.dans.ingest.core.exception.DepositorValidatorException;
import nl.knaw.dans.ingest.core.exception.FailedDepositException;
import nl.knaw.dans.ingest.core.exception.InvalidDatasetStateException;
import nl.knaw.dans.ingest.core.exception.InvalidDepositException;
import nl.knaw.dans.ingest.core.exception.RejectedDepositException;
import nl.knaw.dans.ingest.core.exception.TargetBlockedException;
import nl.knaw.dans.ingest.core.sequencing.TargetedTask;
import nl.knaw.dans.ingest.core.service.mapper.DepositToDvDatasetMetadataMapper;
import nl.knaw.dans.ingest.core.service.mapper.DepositToDvDatasetMetadataMapperFactory;
import nl.knaw.dans.ingest.core.validation.DepositorAuthorizationValidator;
import nl.knaw.dans.lib.dataverse.DataverseException;
import nl.knaw.dans.lib.dataverse.model.dataset.Dataset;
import nl.knaw.dans.lib.dataverse.model.user.AuthenticatedUser;
import nl.knaw.dans.validatedansbag.client.api.ValidateCommandDto;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.nio.file.Path;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class DepositIngestTask implements TargetedTask, Comparable<DepositIngestTask> {

    private static final Logger log = LoggerFactory.getLogger(DepositIngestTask.class);
    protected final String depositorRole;
    protected final Pattern fileExclusionPattern;
    protected final ZipFileHandler zipFileHandler;
    protected final List<URI> supportedLicenses;
    protected final DansBagValidator dansBagValidator;
    protected final DepositToDvDatasetMetadataMapperFactory datasetMetadataMapperFactory;
    protected final Path outboxDir;
    protected final DepositLocation depositLocation;
    protected final DatasetService datasetService;
    private final EventWriter eventWriter;

    private final DepositManager depositManager;
    private final BlockedTargetService blockedTargetService;

    private final DepositorAuthorizationValidator depositorAuthorizationValidator;
    protected Deposit deposit;

    private final String vaultMetadataKey;

    protected boolean deleteDraftOnFailure;

    public DepositIngestTask(
            DepositToDvDatasetMetadataMapperFactory datasetMetadataMapperFactory,
            DepositLocation depositLocation,
            String depositorRole,
            Pattern fileExclusionPattern,
            ZipFileHandler zipFileHandler,
            List<URI> supportedLicenses,
            DansBagValidator dansBagValidator,
            Path outboxDir,
            EventWriter eventWriter,
            DepositManager depositManager,
            DatasetService datasetService,
            BlockedTargetService blockedTargetService,
            DepositorAuthorizationValidator depositorAuthorizationValidator,
            String vaultMetadataKey,
            boolean deleteDraftOnFailure
    ) {
        this.datasetMetadataMapperFactory = datasetMetadataMapperFactory;
        this.depositorRole = depositorRole;
        this.fileExclusionPattern = fileExclusionPattern;
        this.zipFileHandler = zipFileHandler;
        this.supportedLicenses = supportedLicenses;
        this.dansBagValidator = dansBagValidator;
        this.outboxDir = outboxDir;
        this.eventWriter = eventWriter;
        this.depositManager = depositManager;
        this.blockedTargetService = blockedTargetService;
        this.depositLocation = depositLocation;
        this.datasetService = datasetService;
        this.depositorAuthorizationValidator = depositorAuthorizationValidator;
        this.vaultMetadataKey = vaultMetadataKey;
        this.deleteDraftOnFailure = deleteDraftOnFailure;
    }

    public Deposit getDeposit() {
        return this.deposit;
    }

    @Override
    public void run() {
        log.info("START processing deposit {}", depositLocation.getDepositId());
        writeEvent(TaskEvent.EventType.START_PROCESSING, TaskEvent.Result.OK, null);

        // TODO this is really ugly, fix it at some point
        try {
            this.deposit = depositManager.readDeposit(depositLocation);
            log.info("Deposit {} is update: {}", deposit.getDepositId(), deposit.isUpdate());
        } catch (InvalidDepositException e) {
            try {
                updateDepositFromResult(DepositState.FAILED, e.getMessage());
                moveDepositToOutbox(depositLocation.getDir(), OutboxSubDir.FAILED);
            } catch (IOException ex) {
                log.error("Unable to move deposit directory to 'failed' outbox", ex);
            }

            writeEvent(TaskEvent.EventType.END_PROCESSING, Result.FAILED, e.getMessage());
            return;
        }

        try {
            boolean published = doRun();
            if (published) {
                updateDepositFromResult(DepositState.PUBLISHED, "The deposit was successfully ingested in the Data Station and will be automatically archived.");
            } else {
                updateDepositFromResult(DepositState.ACCEPTED,
                        "The deposit was successfully submitted for review in the Data Station and will be reviewed by a data manager before it is published and archived.");
            }
            log.info("END processing (SUCCESS) deposit {}", deposit.getDepositId());
            writeEvent(TaskEvent.EventType.END_PROCESSING, TaskEvent.Result.OK, null);
        } catch (RejectedDepositException e) {
            log.error("END processing (REJECTED) deposit {}", deposit.getDepositId(), e);
            updateDepositFromResult(DepositState.REJECTED, e.getMessage());
            blockTarget(e.getMessage(), DepositState.REJECTED);
            writeEvent(TaskEvent.EventType.END_PROCESSING, TaskEvent.Result.REJECTED, e.getMessage());
        } catch (TargetBlockedException e) {
            log.error("END processing (REJECTED - TARGET BLOCKED) deposit {}", deposit.getDepositId(), e);
            updateDepositFromResult(DepositState.FAILED, e.getMessage());
            writeEvent(TaskEvent.EventType.END_PROCESSING, TaskEvent.Result.FAILED, e.getMessage());
        } catch (Throwable e) {
            log.error("END processing (FAILED) deposit {}", deposit.getDepositId(), e);
            updateDepositFromResult(DepositState.FAILED, e.getMessage());
            blockTarget(e.getMessage(), DepositState.FAILED);
            writeEvent(TaskEvent.EventType.END_PROCESSING, TaskEvent.Result.FAILED, e.getMessage());
        }
    }

    void moveDepositToOutbox(Path path, OutboxSubDir subDir) throws IOException {
        var target = this.outboxDir.resolve(subDir.getValue());
        depositManager.moveDeposit(path, target);
    }

    void updateDepositFromResult(DepositState depositState, String message) {
        deposit.setState(depositState);
        deposit.setStateDescription(message);

        try {
            depositManager.updateAndMoveDeposit(deposit, getTargetPath(depositState));
        } catch (IOException e) {
            log.error("Unable to move directory for deposit {}", deposit.getDir(), e);
        } catch (InvalidDepositException e) {
            log.error("Unable to save deposit for deposit {}", deposit.getDir(), e);
        }
    }

    Path getTargetPath(DepositState depositState) {
        switch (depositState) {
            case PUBLISHED:
            case ACCEPTED:
                return this.outboxDir.resolve(OutboxSubDir.PROCESSED.getValue());
            case REJECTED:
                return this.outboxDir.resolve(OutboxSubDir.REJECTED.getValue());
            case FAILED:
                return this.outboxDir.resolve(OutboxSubDir.FAILED.getValue());
            default:
                throw new IllegalArgumentException(String.format(
                        "Unexpected deposit state '%s' found; not sure where to move it",
                        depositState
                ));

        }
    }

    @Override
    public String getTarget() {
        return depositLocation.getTarget();
    }

    @Override
    public Path getDepositPath() {
        return depositLocation.getDir();
    }

    @Override
    public void writeEvent(TaskEvent.EventType eventType, TaskEvent.Result result, String message) {
        eventWriter.write(getDepositId(), eventType, result, message);
    }

    private UUID getDepositId() {
        return UUID.fromString(depositLocation.getDepositId());
    }

    boolean doRun() throws Exception {
        var deposit = getDeposit();
        var isUpdate = deposit.isUpdate();
        log.debug("Is update: {}", isUpdate);
        if (isUpdate) {
            log.debug("Figuring out the doi for deposit {} ...", deposit.getDepositId());
            var dataverseDoi = resolveDoi(deposit);
            log.debug("Found target DOI for deposit {}: {}", deposit.getDepositId(), dataverseDoi);
            deposit.setDataverseDoi(dataverseDoi);
            log.debug("Checking if dataset {} is allowed to be updated by user {} ...", dataverseDoi, deposit.getDepositorUserId());
            if (!isDatasetUpdateAllowed()) {
                throw new RejectedDepositException(deposit, String.format(
                        "Dataset %s is not allowed to be updated by user %s", deposit.getDataverseDoi(), deposit.getDepositorUserId()
                ));
            }
            log.debug("Checking if dataset {} is in review ...", dataverseDoi);
            if (isDatasetInReview()) {
                throw new RejectedDepositException(deposit, String.format(
                        "Dataset %s is in review and cannot be updated", deposit.getDataverseDoi()
                ));
            }
            log.debug("Checking if dataset {} is blocked ...", dataverseDoi);
            checkBlockedTarget();
        }

        checkDoiRequirements();
        validateDeposit();
        return createOrUpdateDataset(isUpdate);
    }

    boolean createOrUpdateDataset(boolean isUpdate) throws Exception {
        var dataverseDataset = getMetadata();
        var persistentId = isUpdate
                ? newDatasetUpdater(dataverseDataset).performEdit()
                : newDatasetCreator(dataverseDataset, depositorRole).performEdit();

        if (isDatasetPublicationAllowed()) {
            publishDataset(persistentId);
            log.debug("Dataset {} published", persistentId);
            postPublication(persistentId);
            return true;
        } else {
            submitForReview(persistentId);
            log.debug("Dataset {} submitted for review", persistentId);
            postSubmitForReview(persistentId);
            return false;
        }
    }

    private boolean isDatasetPublicationAllowed() {
        try {
            return depositorAuthorizationValidator.isDatasetPublicationAllowed(deposit);
        } catch (DepositorValidatorException e) {
            throw new FailedDepositException(deposit, e.getMessage());
        }
    }

    private boolean isDatasetUpdateAllowed() {
        try {
            return depositorAuthorizationValidator.isDatasetUpdateAllowed(deposit);
        } catch (DepositorValidatorException e) {
            throw new FailedDepositException(deposit, e.getMessage());
        }
    }

    boolean isDatasetInReview() throws IOException, DataverseException {
        var deposit = getDeposit();
        var target = deposit.getDataverseDoi();
        return datasetService.isDatasetInReview(target);
    }

    void checkBlockedTarget() throws TargetBlockedException {
        var deposit = getDeposit();
        var target = deposit.getDataverseDoi();
        var depositId = deposit.getDepositId();
        if (blockedTargetService.isBlocked(target)) {
            throw new TargetBlockedException(String.format(
                    "Deposit with id %s and target %s is blocked by a previous deposit", depositId, target
            ));
        }
    }

    void blockTarget(String message, DepositState depositState) {
        var deposit = getDeposit();
        var target = deposit.getDataverseDoi();
        // TODO: Shouldn't target be a sword token?

        if (target == null) {
            log.debug("Target for deposit {} is null, unable to block target. This probably means it is the first version of a deposit and can be ignored", deposit.getDepositId());
            return;
        }

        try {
            blockedTargetService.blockTarget(
                    deposit.getDepositId(),
                    target,
                    depositState.toString(),
                    message
            );
        } catch (TargetBlockedException e) {
            log.warn("Target {} is already blocked", target);
        }
    }

    void checkDoiRequirements() {
        var hasDoi = StringUtils.isNotBlank(deposit.getDoi());

        if (hasDoi) {
            throw new IllegalArgumentException("Deposits must not have an identifier.doi property unless they are migrated");
        }
    }

    void validateDeposit() {
        var result = dansBagValidator.validateBag(
                deposit.getBagDir(), ValidateCommandDto.PackageTypeEnum.DEPOSIT);

        if (!result.getIsCompliant()) {
            var violations = result.getRuleViolations().stream()
                    .map(r -> String.format("- [%s] %s", r.getRule(), r.getViolation()))
                    .collect(Collectors.joining("\n"));

            throw new RejectedDepositException(deposit, String.format(
                    "Bag was not valid according to Profile Version %s. Violations: %s",
                    result.getProfileVersion(), violations)
            );
        }
    }

    void postSubmitForReview(String persistentId) throws IOException, DataverseException, InterruptedException {
        saveDoiInDepositProperties(persistentId);
    }

    void postPublication(String persistentId) throws IOException, DataverseException, InterruptedException {
        try {
            datasetService.waitForState(persistentId, "RELEASED");
            saveDoiInDepositProperties(persistentId);
            saveUrnInDepositProperties(persistentId);
        } catch (InvalidDatasetStateException e) {
            throw new FailedDepositException(deposit, e.getMessage());
        }
    }

    // Does not actually save the DOI, but only sets it on the deposit object
    void saveDoiInDepositProperties(String persistentId) throws IOException, DataverseException {
        var basePersistentId = persistentId;
        if (persistentId.startsWith("doi:")) {
            basePersistentId = persistentId.substring("doi:".length());
        }

        deposit.setDoi(basePersistentId);
    }

    // Does not actually save the URN, but only sets it on the deposit object
    void saveUrnInDepositProperties(String persistentId) throws IOException, DataverseException {
        var urn = datasetService.getDatasetUrnNbn(persistentId)
                .orElseThrow(() -> new IllegalStateException(String.format("Dataset %s did not obtain a URN:NBN", persistentId)));

        deposit.setUrn(urn);
    }


    void publishDataset(String persistentId) throws Exception {
        try {
            datasetService.publishDataset(persistentId);
        } catch (IOException | DataverseException e) {
            log.error("Unable to publish dataset", e);
            throw e;
        }
    }

    void submitForReview(String persitentId) throws Exception {
        datasetService.submitForReview(persitentId);
    }

    DatasetEditor newDatasetUpdater(Dataset dataset, boolean isMigration, boolean deleteDraftOnFailure) {
        return new DatasetUpdater(
                isMigration,
                dataset,
                deposit,
                supportedLicenses,
                fileExclusionPattern,
                zipFileHandler,
                new ObjectMapper(),
                datasetService,
                vaultMetadataKey,
                deleteDraftOnFailure
        );
    }

    DatasetEditor newDatasetUpdater(Dataset dataset) {
        return newDatasetUpdater(dataset, false, false);
    }

    DatasetEditor newDatasetCreator(Dataset dataset, String depositorRole, boolean isMigration) {
        return new DatasetCreator(
                isMigration,
                dataset,
                deposit,
                new ObjectMapper(),
                supportedLicenses,
                fileExclusionPattern,
                zipFileHandler,
                depositorRole,
                datasetService,
                vaultMetadataKey,
                deleteDraftOnFailure
        );
    }

    DepositToDvDatasetMetadataMapper newMapper() {
        return datasetMetadataMapperFactory.createMapper(false, false);
    }

    DatasetEditor newDatasetCreator(Dataset dataset, String depositorRole) {
        return newDatasetCreator(dataset, depositorRole, false);
    }

    Dataset getMetadata() {
        var date = getDateOfDeposit();
        var contact = getDatasetContact();
        var mapper = newMapper();
        return mapper.toDataverseDataset(
                deposit.getDdm(),
                deposit.getOtherDoiId(),
                date.orElse(null),
                contact.orElse(null),
                deposit.getVaultMetadata(),
                deposit.getDepositorUserId(),
                deposit.restrictedFilesPresent(),
                deposit.getHasOrganizationalIdentifier(),
                deposit.getHasOrganizationalIdentifierVersion()
        );
    }

    Optional<String> getDateOfDeposit() {
        return Optional.empty();
    }

    Optional<AuthenticatedUser> getDatasetContact() {
        return Optional.ofNullable(deposit.getDepositorUserId())
                .filter(StringUtils::isNotBlank)
                .map(userId -> datasetService.getUserById(userId)
                        .orElseThrow(() -> new RuntimeException("Unable to fetch user with id " + userId)));
    }

    @Override
    public int compareTo(DepositIngestTask depositIngestTask) {
        return getCreatedInstant().compareTo(depositIngestTask.getCreatedInstant());
    }

    public OffsetDateTime getCreatedInstant() {
        return depositLocation.getCreated();
    }

    String resolveDoi(Deposit deposit) throws IOException, DataverseException {
        return getDoi("dansSwordToken", deposit.getVaultMetadata().getSwordToken());
    }

    String getDoi(String key, String value) throws IOException, DataverseException {
        var items = datasetService.searchDatasets(key, value);

        if (items.size() != 1) {
            throw new FailedDepositException(deposit, String.format(
                    "Deposit is update of %s datasets; should always be 1!", items.size()
            ), null);
        }

        var doi = items.get(0).getGlobalId();
        log.debug("Deposit is update of dataset {}", doi);
        return doi;
    }
}
